File: facebook-common/src/main/java/com/facebook/login/KatanaProxyLoginMethodHandler.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.login;

import android.content.Intent;
import android.os.Parcel;
import android.os.Parcelable;
import com.facebook.FacebookSdk;
import com.facebook.internal.CustomTabUtils;
import com.facebook.internal.NativeProtocol;
import com.facebook.internal.ServerProtocol;
import java.util.List;

class KatanaProxyLoginMethodHandler extends NativeAppLoginMethodHandler {

  KatanaProxyLoginMethodHandler(LoginClient loginClient) {
    super(loginClient);
  }

  @Override
  String getNameForLogging() {
    return "katana_proxy_auth";
  }

  @Override
  int tryAuthorize(LoginClient.Request request) {
    LoginBehavior behavior = request.getLoginBehavior();
    boolean ignoreAppSwitchToLoggedOut =
        FacebookSdk.ignoreAppSwitchToLoggedOut
            && CustomTabUtils.getChromePackage() != null
            && behavior.allowsCustomTabAuth();
    String e2e = LoginClient.getE2E();
    List<Intent> intents =
        NativeProtocol.createProxyAuthIntents(
            loginClient.getActivity(),
            request.getApplicationId(),
            request.getPermissions(),
            e2e,
            request.isRerequest(),
            request.hasPublishPermission(),
            request.getDefaultAudience(),
            getClientState(request.getAuthId()),
            request.getAuthType(),
            ignoreAppSwitchToLoggedOut,
            request.getMessengerPageId(),
            request.getResetMessengerState(),
            request.isFamilyLogin(),
            request.shouldSkipAccountDeduplication(),
            request.getNonce());

    addLoggingExtra(ServerProtocol.DIALOG_PARAM_E2E, e2e);

    if (intents == null) {
      return 0;
    }

    for (int i = 0; i < intents.size(); i++) {
      boolean launchedIntent = tryIntent(intents.get(i), LoginClient.getLoginRequestCode());
      if (launchedIntent) {
        return i + 1;
      }
    }

    return 0;
  }

  @Override
  public boolean shouldKeepTrackOfMultipleIntents() {
    return true;
  }

  KatanaProxyLoginMethodHandler(Parcel source) {
    super(source);
  }

  @Override
  public int describeContents() {
    return 0;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    super.writeToParcel(dest, flags);
  }

  public static final Parcelable.Creator<KatanaProxyLoginMethodHandler> CREATOR =
      new Parcelable.Creator<KatanaProxyLoginMethodHandler>() {

        @Override
        public KatanaProxyLoginMethodHandler createFromParcel(Parcel source) {
          return new KatanaProxyLoginMethodHandler(source);
        }

        @Override
        public KatanaProxyLoginMethodHandler[] newArray(int size) {
          return new KatanaProxyLoginMethodHandler[size];
        }
      };
}

File: facebook-common/src/main/java/com/facebook/login/KatanaProxyLoginMethodHandler.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.login

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.VisibleForTesting
import com.facebook.FacebookSdk
import com.facebook.internal.CustomTabUtils.getChromePackage
import com.facebook.internal.NativeProtocol.createProxyAuthIntents
import com.facebook.internal.ServerProtocol

@VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
class KatanaProxyLoginMethodHandler : NativeAppLoginMethodHandler {
  constructor(loginClient: LoginClient) : super(loginClient)

  override val nameForLogging = "katana_proxy_auth"

  override fun tryAuthorize(request: LoginClient.Request): Int {
    val behavior = request.loginBehavior
    val ignoreAppSwitchToLoggedOut =
        (FacebookSdk.ignoreAppSwitchToLoggedOut &&
            getChromePackage() != null &&
            behavior.allowsCustomTabAuth())
    val e2e = LoginClient.getE2E()
    val intents =
        createProxyAuthIntents(
            loginClient.activity,
            request.applicationId,
            request.permissions,
            e2e,
            request.isRerequest,
            request.hasPublishPermission(),
            request.defaultAudience ?: DefaultAudience.NONE,
            getClientState(request.authId),
            request.authType,
            ignoreAppSwitchToLoggedOut,
            request.messengerPageId,
            request.resetMessengerState,
            request.isFamilyLogin,
            request.shouldSkipAccountDeduplication(),
            request.nonce,
            request.codeChallenge,
            request.codeChallengeMethod?.name)
    addLoggingExtra(ServerProtocol.DIALOG_PARAM_E2E, e2e)
    for ((i, intent) in intents.withIndex()) {
      val launchedIntent = tryIntent(intent, LoginClient.getLoginRequestCode())
      if (launchedIntent) {
        return i + 1
      }
    }
    return 0
  }

  override fun shouldKeepTrackOfMultipleIntents(): Boolean = true

  constructor(source: Parcel) : super(source)

  override fun describeContents(): Int = 0

  companion object {
    @JvmField
    val CREATOR: Parcelable.Creator<KatanaProxyLoginMethodHandler> =
        object : Parcelable.Creator<KatanaProxyLoginMethodHandler> {
          override fun createFromParcel(source: Parcel): KatanaProxyLoginMethodHandler {
            return KatanaProxyLoginMethodHandler(source)
          }

          override fun newArray(size: Int): Array<KatanaProxyLoginMethodHandler?> {
            return arrayOfNulls(size)
          }
        }
  }
}
